src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -278,6 +285,11 @@ private void WatchForFileSystemEventsThreadStart(object arg)
                 {
                     // Always unschedule the RunLoop before cleaning up
                     Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+
+                    // Release the WatcherLoop Core Foundation object.
+                    CFStringRef release = _watcherRunLoop;
+                    _watcherRunLoop = IntPtr.Zero;
+                    Interop.CoreFoundation.CFRelease(release);
```

Comment:
I couldn't decide whether that race condition was likely enough to guard against. It's significantly rarer than the scenario that was causing the intermittent failures, but it can potentially still cause the segfault. Better safe than sorry when it comes to segfaults though; I pushed a fix.